velox/common/memory/MemoryUsageTracker.cpp

```diff
@@ -33,9 +33,74 @@ std::shared_ptr<MemoryUsageTracker> MemoryUsageTracker::create(
   return std::make_shared<SharedMemoryUsageTracker>(parent, type, config);
 }
 
+void MemoryUsageTracker::update(int64_t size, bool /* mock */) {
+  if (size > 0) {
+    int64_t increment = 0;
+    ++usage(numAllocs_, UsageType::kTotalMem);
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      if (usedReservation_ + size > reservation_) {
+        increment = reserveLocked(size);
+      }
+    }
+    checkAndPropagateReservationIncrement(increment, false);
+    usedReservation_.fetch_add(size);
+    return;
+  }
+  // Decreasing usage. See if need to propagate upward.
+  int64_t decrement = 0;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    auto newUsed = usedReservation_ += size;
+    auto newCap = std::max(minReservation_.load(), newUsed);
+    auto newQuantized = quantizedSize(newCap);
+    if (newQuantized != reservation_) {
+      decrement = reservation_ - newQuantized;
+      reservation_ = newQuantized;
+    }
+  }
+  if (decrement) {
+    decrementUsage(type_, decrement);
+  }
+}
+
+void MemoryUsageTracker::reserve(int64_t size) {
+  int64_t increment;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    increment = reserveLocked(size);
+    minReservation_ = reservation_.load();
```

Comment:
nit: we don't need to call load() here? Default read access mode should be sufficient?